drivers/memory_window/dynamic_slots.c

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "dynamic_slots.h"
+
+#include <zephyr/drivers/mm/system_mm.h>
+#include <zephyr/arch/xtensa/cache.h>
+
+int dynamic_slots_init(const struct device *dev)
+{
+	uint32_t buf_size =  MEMORY_WINDOW_SLOT_SIZE * MEMORY_WINDOW_SLOTS_COUNT;
+	const struct mem_win_config *config = dev->config;
+
+	uint32_t *virt_buf = (uint32_t *)
+		arch_xtensa_cached_ptr((void *)(config->mem_base));
+
+	uint32_t err = sys_mm_drv_map_region(virt_buf+(ADSP_PAGE_SIZE/sizeof(uint32_t)), 0,
+			buf_size, SYS_MM_MEM_CACHE_WB);
+
+	if (err != ADSP_SUCCESS)
+		return err;
+
+	memset(virt_buf, 0, (buf_size+ADSP_PAGE_SIZE));
+	z_xtensa_cache_flush_inv(virt_buf, (buf_size+ADSP_PAGE_SIZE));
+
+	return ADSP_SUCCESS;
+}
+
+int dynamic_slots_map_slot(const struct device *dev, uint32_t resource_id,
+		void *virtual_memory_address, uint32_t slot_id)
+{
+#if CONFIG_MM_DRV_INTEL_ADSP_MTL_TLB
+	const struct mem_win_config *config = dev->config;
+
+	uint32_t *buf = (uint32_t *)
+		arch_xtensa_cached_ptr((void *)(config->mem_base));
+	uint32_t buf_size =
+		ADSP_PAGE_SIZE + MEMORY_WINDOW_SLOT_SIZE*MEMORY_WINDOW_SLOTS_COUNT;
+
+	z_xtensa_cache_flush_inv(buf, buf_size);
+
+	struct MemoryWindow *memory_window = (struct MemoryWindow *)(z_soc_uncached_ptr(buf));
+
+	size_t first_free = ARRAY_SIZE(memory_window->slot_descs);
+
+	for (size_t idx = 0; idx < ARRAY_SIZE(memory_window->slot_descs); ++idx) {
+		if (memory_window->slot_descs[idx].vma == (uint32_t)virtual_memory_address &&
+			memory_window->slot_descs[idx].resource_id == resource_id &&
+			memory_window->slot_descs[idx].slot_id == slot_id) {
+			return ADSP_SUCCESS;
+		}
+		if (first_free == ARRAY_SIZE(memory_window->slot_descs) &&
+			memory_window->slot_descs[idx].vma == 0) {
+			first_free = idx;
+		}
+	}
+
+	if (first_free == ARRAY_SIZE(memory_window->slot_descs))
+		return ADSP_NOT_FOUND;
```

Comment:
I think internal infrastructure should use POSIX error codes. Only IPC responses for IPC4 should use positive error codes.
This applies for entire commit.